codec-http/src/main/java/io/netty/handler/codec/http/HttpServerKeepAliveHandler.java

```diff
@@ -0,0 +1,127 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http;
+
+import io.netty.channel.ChannelDuplexHandler;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPipeline;
+import io.netty.channel.ChannelPromise;
+
+import static io.netty.handler.codec.http.HttpUtil.*;
+
+/**
+ * HttpServerKeepAliveHandler helps close persistent connections when appropriate.
+ * <p>
+ * The server channel is expected to set the proper 'Connection' header if it can handle persistent connections. {@link
+ * HttpServerKeepAliveHandler} will automatically close the channel for any LastHttpContent that corresponds to a client
+ * request for closing the connection, or if the HttpResponse associated with that LastHttpContent requested closing the
+ * connection or didn't have a self defined message length.
+ * <p>
+ * Since {@link HttpServerKeepAliveHandler} expects {@link HttpObject}s it should be added after {@link HttpServerCodec}
+ * but before any other handlers that might send a {@link HttpResponse}. <blockquote>
+ * <pre>
+ *  {@link ChannelPipeline} p = ...;
+ *  ...
+ *  p.addLast("serverCodec", new {@link HttpServerCodec}());
+ *  p.addLast("httpKeepAlive", <b>new {@link HttpServerKeepAliveHandler}()</b>);
+ *  p.addLast("aggregator", new {@link HttpObjectAggregator}(1048576));
+ *  ...
+ *  p.addLast("handler", new HttpRequestHandler());
+ *  </pre>
+ * </blockquote>
+ */
+public class HttpServerKeepAliveHandler extends ChannelDuplexHandler {
+    private static final String MULTIPART_PREFIX = "multipart";
+
+    private boolean persistentConnection = true;
+    // Track pending responses to support client pipelining: https://tools.ietf.org/html/rfc7230#section-6.3.2
+    private int pendingResponses;
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        // read message and track if it was keepAlive
+        if (msg instanceof HttpRequest) {
+            final HttpRequest request = (HttpRequest) msg;
+            if (persistentConnection) {
+                pendingResponses += 1;
+                persistentConnection = isKeepAlive(request);
+            }
+        }
+        super.channelRead(ctx, msg);
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        // modify message on way out to add headers if needed
+        if (msg instanceof HttpResponse) {
+            final HttpResponse response = (HttpResponse) msg;
+            trackResponse(response);
+            // Assume the response writer knows if they can persist or not and sets isKeepAlive on the response
+            if (!isKeepAlive(response) || !isSelfDefinedMessageLength(response)) {
```

Comment:
https://tools.ietf.org/html/rfc7230#section-6.3
>HTTP/1.1 defaults to the use of "persistent connections"

Does this statement account for that behavior?